### DISCUSSION TITLE
Workarounds when facing an issue?

### DISCUSSION THREAD

**Discussion by rhinot** [keepassxreboot/keepassxc#10302] [Q&A]
Hey All,
Been using KeePassXC a long time and love it. I'm currently [facing an issue](https://github.com/keepassxreboot/keepassxc/issues/8368) that prevents me from saving, which means KPXC is useless for me for any new accounts/passwords.

I know the Devs have many things to fix, so this may not get fixed right away. Does anyone have any workarounds that I could use to store passwords temporarily that isn't completely unsafe?

**Comment by droidmonkey** [comment]
Possibly create a new database or change to a different storage method.

**Reply by rhinot** [reply]
Thanks!

I tried all storage methods. All fail sadly.

Is the idea that I start a new database and store the new passwords until the issue is fixed, or that I could potentially import passwords in from the old database?

**Reply by droidmonkey** [reply]
It is totally unclear why this issue is happening to a very small set of users, so it is hard to fix it. Any insights into why you have perpetual failures with saving a file? We are not doing anything fancy.

**Comment by rhinot** [comment]
It could be something about it being a mounted network drive, although this
hasn't been an issue in the past - even when the drive went offline.
My guess is that some state was set or threshold met within keepass, that's
causing this to fail. For example, it could be that the kdbx file reached a
certain size and thus a timeout limit is hit.

I haven't looked at the code to know what variables are set/monitored on a
write. I'm guessing you know well though. :)

Are there any similarities between the different users who are experiencing
the same issue? Are they all on the same OS? Networked drives? Certain size
KDBX files?

I'm happy to run a debug build, if you think it would be helpful to
understand what's causing this.

**Reply by droidmonkey** [reply]
It seems to be generally network drives. Did you change your database encryption settings at all?

Basically, the sequence that causes that error message requires a database to get stuck in a save operation, like an infinite loop. Either saving is taking forever and not timing out, or the key derivation is not completing. 

This could also be a deadlock situation where you save and immediately the database is attempting to lock or close which might cause the mutex check to deadlock.

If you enable alternative saving method - direct write save, does this still happen? https://keepassxc.org/docs/KeePassXC_UserGuide#_advanced_save_options

**Reply by PredatorIWD** [reply]
In regards to mutex and a possible cause of this problem in my case that seemingly randomly stopped happening that [I mentioned in the original issue thread](https://github.com/keepassxreboot/keepassxc/issues/8368#issuecomment-1483581683), it wasn't a network drive but I did have the database file be inside a folder on the OS drive that was synced with the Dropbox desktop app to the Dropbox servers by me launching that app every few weeks, syncing and then closing it completely, at a similar interval at which this problem was happening.

Looking back at it now, the problem not happening ever again does fit the timeline of around the time when I stopped using Dropbox altogether, especially given the fact that it stopped happening without me changing the KeePassXC version or my hardware/storage setup.

**Comment by rhinot** [comment]
I didn't, but that's super helpful.

I'd think it was the mutex check, but this happened independent of a
product update, so I'm guessing this wouldn't just randomly start one day.

Is it possible the key was corrupted that prevents saves, but allows opens?
If so, is there a way I could check for that?

**Reply by droidmonkey** [reply]
The key is cycled (recreated) on every save, that wouldn't be the cause. This is most likely a file operation gone sideways. Something is causing the file save to enter an infinite timeout.

**Reply by rhinot** [reply]
I ended up exporting to CSV, then imported to create a new database. Tried that file locally and on the mounted network drive. No issues.

Sounds like it is some type of file-level corruption.

**Reply by droidmonkey** [reply]
The file is recreated entirely on every save. Perhaps there is an infinite loop happening because of something with your entries, and a save can never finish. In other words, the file itself is not the issue. Something in how we process data in your database is.
